### PR TITLE
Feat/memtable freeze

### PR DESCRIPTION
**Pull request by KKould**
Implement `Db::freeze` to convert elements in `Memtable` into Arrow RecordBatch and push them in immutable

### CODE REVIEW

File: src/lib.rs [hunk 1]
```diff
@@ -190,6 +195,42 @@ where
...
+            Field::new("key", DataType::LargeBinary, false),
+            Field::new("value", DataType::LargeBinary, true),
+        ]));
+        let mut buf = Cursor::new(vec![0; 128]);
+        let mut key_builder = GenericBinaryBuilder::<Offset>::new();
+        let mut value_builder = GenericBinaryBuilder::<Offset>::new();
+
+        for (key, value) in mem_table.iter() {
+            clear(&mut buf);
+            key.encode(&mut buf).await.unwrap();
```
Comment (KKould) [inline src/lib.rs:218]: TODO: The encoded error has not been processed for the time being.

File: src/lib.rs [hunk 2]
```diff
@@ -190,6 +196,43 @@ where
...
+        let mut value_builder = GenericBinaryBuilder::<Offset>::new();
+
+        for (key, value) in mem_table.iter() {
+            clear(&mut buf);
+            key.encode(&mut buf).await.map_err(EncodeError::Key)?;
+            key_builder.append_value(buf.get_ref());
+
+            if let Some(value) = value {
+                clear(&mut buf);
+                value.encode(&mut buf).await.unwrap();
```
Comment (KKould) [inline src/lib.rs:224]: I don't know how to handle this error

File: Cargo.toml [hunk 3]
```diff
@@ -13,6 +13,7 @@ crossbeam-queue = "0.3"
 executor = { git = "https://github.com/ethe/executor.git", branch = "main" }
 futures = "0.3"
 fxhash = "0.2"
+lazy_static = "1.4.0"
```
Comment (ethe) [inline Cargo.toml:16]: it should be better to specify major version of a stabled crate `lazy_static = "1"`

File: src/lib.rs [hunk 4]
```diff
@@ -47,7 +63,7 @@ where
     wal_manager: Arc<WalManager<WP>>,
     pub(crate) mutable_shards:
         Shard<unsend::lock::RwLock<MutableShard<K, V, O::Timestamp, WP::File>>>,
-    pub(crate) immutable: ArrayQueue<MemTable<K, V, O::Timestamp>>,
+    pub(crate) immutable: ArrayQueue<RecordBatch>,
```
Comment (ethe) [inline src/lib.rs:66]: immutable tables have to keep its WAL, it needs to be recovered from WAL once DB shutdowns before

**Comment by ethe** [comment]
Also I think we might need to rebuild the index of immutable tables: `Arc<BtreeMap<InternalKey<K>, u32>>`, its value points to the offset of the record in record batch. It is created in two cases:
1. freezing from mutable memtable,
2. recover from WAL,

In current, build arrow format immutable table and its index is expensive, we could adjust mutable memtable layout later, it could transmute mutable memtable to immutable very cheap.

File: .github/workflows/ci.yml [hunk 5]
```diff
@@ -0,0 +1,67 @@
...
+          components: rustfmt, clippy
+
+      # `cargo check` command here will use installed `nightly`
+      # as it is set as an "override" for current directory
+
+      - name: Run cargo fmt
+        uses: actions-rs/cargo@v1
+        with:
+          command: fmt
+          args: -- --check
```
Comment (ethe) [inline .github/workflows/ci.yml:67]: it should have a new line at the end of files

File: Cargo.toml [hunk 6]
```diff
@@ -4,6 +4,7 @@ name = "elsm"
 version = "0.1.0"
 
 [dependencies]
+arrow = "51.0.0"
```
Comment (ethe) [inline Cargo.toml:7]: also `arrow="51"`

File: src/lib.rs [hunk 7]
```diff
@@ -167,6 +189,74 @@ where
             .await
     }
 
+    async fn range_scan<G>(
```
Comment (ethe) [inline src/lib.rs:192]: I think freezing may not blocked range_scan implementation? It should be better in another PR

File: src/lib.rs (hunk 4 above)
Comment (KKould) [reply src/lib.rs:66]: I think this can be handled in another pr

issue: https://github.com/from-the-basement/elsm/issues/8

File: src/lib.rs (hunk 2 above)
Comment (KKould) [reply src/lib.rs:229]: unable to use `Box<dyn Error>`, this results in Send + Sync not being possible

**Comment by KKould** [comment]
> Also I think we might need to rebuild the index of immutable tables: `Arc<BtreeMap<InternalKey<K>, u32>>`, its value points to the offset of the record in record batch. It is created in two cases:
> 
> 1. freezing from mutable memtable,
> 2. recovering from WAL,
> 
> In current, build arrow format immutable table and its index is expensive, we could adjust mutable memtable layout later, it could transmute mutable memtable to immutable very cheap.

I retained the `InternalKey` and encapsulated `RecordBatch` into `IndexBatch`. then we can use `IndexBatch::find` to query the value.

**Review by ethe — APPROVED** [review]
